[MRG] MNT Use `clip` to make dictionary and/or code positive

jakirkham opened the pull request:
Instead of finding a mask of negative values and setting them to zero, use NumPy's `clip` to replace all values less than zero with zero. These have the same effect other than `clip` is a bit faster as it avoids creating an intermediate bool array for setting values instead doing this in-place.

<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->


#### What does this implement/fix? Explain your changes.


#### Any other comments?


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

jakirkham committed:
Use `clip` to make dictionary positive

Instead of finding a mask of negative values and setting them to zero,
use NumPy's `clip` to replace all values less than zero with zero. These
have the same effect other than `clip` is a bit faster as it avoids
creating an intermediate bool array for setting values instead doing
this in-place.

rth reviewed (approved):
Thanks @jakirkham ! LGTM, assuming tests pass.

Does this have a measurable impact on performance? (Not asking for benchmarks, just curious..)

rth commented:
BTW, there are a few other places where this could also apply,
```py
sklearn/decomposition/dict_learning.py:163:            new_code[new_code < 0] = 0
sklearn/neighbors/nearest_centroid.py:164:            deviation[deviation < 0] = 0
sklearn/metrics/classification.py:1913:    losses[losses <= 0] = 0
sklearn/gaussian_process/correlation_models.py:241:    td[td > 1.] = 1.
sklearn/gaussian_process/correlation_models.py:293:    td[td > 1.] = 1.
sklearn/impute.py:755:        abs_corr_mat[abs_corr_mat < tolerance] = tolerance
```

The performance impact appears to be not that large,
```py
In [1]: import numpy as np

In [7]: X = np.random.RandomState(42).rand(100000, 200)

In [8]: %%time
   ...: X[X<0] = 0
   ...: 
CPU times: user 18.9 ms, sys: 5.98 ms, total: 24.9 ms
Wall time: 23.7 ms

In [9]: X = np.random.RandomState(42).rand(100000, 200)

In [10]: %%time
    ...: np.clip(X, 0, None, out=X)
    ...: 
CPU times: user 20.6 ms, sys: 0 ns, total: 20.6 ms
Wall time: 20.7 ms

```
but I imagine it will also use less memory since it operates inplace.

jakirkham commented:
So there are a couple things to take into account when benchmarking this. First as this is mutating the original array, subsequent operations on the same array are no-ops. Second things like `rand` don't generate negative numbers, which is what we are removing in this operation.

Have addressed these by writing functions that copy the array to be operated on beforehand and generate random numbers in the interval `-1` to `1`. Of course copying has some overhead as well so created a null case where the array is merely copied so as to remove that impact. Code with some simplistic benchmarks below.

```python
In [1]: import numpy as np

In [2]: a = 2 * np.random.random((100, 200)) - 1

In [3]: def test0(a):
   ...:     a = a.copy()
   ...:     return a

In [4]: def test1(a):
   ...:     a = a.copy()
   ...:     a[:, 3][a[:, 3] < 0] = 0
   ...:     return a

In [5]: def test2(a):
   ...:     a = a.copy()
   ...:     np.clip(a[:, 3], 0, None, out=a[:, 3])
   ...:     return a

In [6]: %timeit test0(a)
10.4 µs ± 471 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [7]: %timeit test1(a)
16.2 µs ± 844 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [8]: %timeit test2(a)
14.9 µs ± 159 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

So on this particular dummy dataset this is looking like a ~30% speed improvement. Have not tested other sizes.

jakirkham committed:
Use `clip` in `_sparse_encode`'s `threshold` case

Instead of finding a mask of negative values and setting them to zero,
use NumPy's `clip` to replace all values less than zero with zero. These
have the same effect other than `clip` is a bit faster as it avoids
creating an intermediate bool array for setting values instead doing
this in-place.

jakirkham commented:
Added on the `code` case as that seemed related. Happy to add the other cases in another PR.

rth commented:
Thanks for the PR and the additional benchmarks @jakirkham !

> Happy to add the other cases in another PR.

That would be appreciated, thanks! I'm not sure other cases are in performance critical parts of the code, but it wouldn't hurt in any case..

jakirkham commented:
Of course. :)

Ok, put together PR ( https://github.com/scikit-learn/scikit-learn/pull/11474 ) to fix cases that matched `git grep '\[\w*.*\w*[<>][ =]\w*.*\w*\] ='` (with some editing as that doesn't guarantee that the operation is thresholding).
